src/Chainweb/Pact/Backend/PactState.hs

```diff
@@ -0,0 +1,337 @@
+diffLatestPactState s1 s2 = do
+  let diff :: UserTable -> UserTable -> UserTableDiff
+      diff ut1 ut2
+        | ut1.tableName /= ut2.tableName = error "diffLatestPactState: mismatched table names"
+        | otherwise = UserTableDiff ut1.tableName
+            $ List.filter (not . PatienceM.isSame)
+            $ Patience.pairItems (\x y -> x.rowKey == y.rowKey)
+            $ Patience.diff ut1.rows ut2.rows
+
+  S.zipWith diff s1 s2
+
+data UserTableDiff = UserTableDiff
+  { tableName :: !Text
+  , rowDiff :: [Delta PactRow]
+  }
+  deriving stock (Eq, Ord, Show)
+
+--instance NFData UserTableDiff where
+--  rnf (UserTableDiff x y) = rnf x `seq` rnf y `seq` ()
+
+instance ToJSON UserTableDiff where
+  toJSON utd = Aeson.object
+    [ "table_name" .= utd.tableName
+    , "row_diff" .= List.map deltaToObject utd.rowDiff
+    ]
+    where
+      deltaToObject :: (ToJSON a) => Delta a -> Aeson.Value
+      deltaToObject = \case
+        Old x -> Aeson.object
+          [ "old" .= x
+          ]
+        New x -> Aeson.object
+          [ "new" .= x
+          ]
+        Delta x y -> Aeson.object
+          [ "old" .= x
+          , "new" .= y
+          ]
+        Same _ -> Aeson.Null
+
+data UserTable = UserTable
+  { tableName :: !Text
+  , rows :: [PactRow]
+  }
+  deriving stock (Eq, Ord, Show)
+
+data PactRow = PactRow
+  { rowKey :: !ByteString
+  , rowData :: !ByteString
+  , txId :: !Int64
+  }
+  deriving stock (Eq, Ord, Show)
+
+instance ToJSON PactRow where
+  toJSON pr = Aeson.object
+    [ "row_key" .= Text.decodeUtf8 pr.rowKey
+    , "row_data" .= Text.decodeUtf8 pr.rowData
+    , "tx_id" .= pr.txId
+    ]
+
+getActiveRows :: UserTable -> UserTable
+getActiveRows (UserTable name rows) = UserTable name
+  $ List.map takeHead
+  $ List.map (List.sortOn (Down . txId))
+  $ List.groupBy (\x y -> rowKey x == rowKey y)
+  $ List.sortOn rowKey rows
+  where
+    takeHead :: [a] -> a
+    takeHead = \case
+      [] -> error "getLatestPactState.getActiveRows.takeHead: impossible case"
+      (x : _) -> x
+
+utf8ToText :: Utf8 -> Text
+utf8ToText (Utf8 u) = Text.decodeUtf8 u
+
+data Config = Config
+  { pactDbDir :: FilePath
+  , chainwebVersion :: ChainwebVersion
+  , logDir :: FilePath
+  , numThreads :: Int
+  }
+
+main :: IO ()
+main = do
+  cfg <- execParser opts
+
+  cids <- do
+    -- Get the latest block height on chain 0 for the purpose of calculating all
+    -- the chain ids at the current (version,height) pair
+    latestBlockHeight <- C.withDefaultLogger Error $ \logger -> do
+      let resetDb = False
+      withSqliteDb (unsafeChainId 0) logger cfg.pactDbDir resetDb $ \(SQLiteEnv db _) -> do
+        getLatestBlockHeight db
+    pure $ List.sort $ F.toList $ chainIdsAt cfg.chainwebVersion latestBlockHeight
+
+  -- Compaction is destructive. In order to do a streaming diff, we need to make
+  -- a copy of the pact state.
+  withSystemTempDirectory "pact-compact-diff" $ \tmpPactDbDir -> do
```

Comment:
there should be an argument for where the copy should go. and there should be a check that it's different from the pactDbDir. this way the post-compaction db can be inspected.